### DISCUSSION TITLE
Ebitengine on raspberry pi?

### DISCUSSION THREAD

**Discussion by irai** [hajimehoshi/ebiten#2841] [Q&A] [answered]
I am trying to run Ebitengine on the raspberry PI 3 with Bookworm OS and get the following GLX error when trying to run the test rotate example (go run github.com/hajimehoshi/ebiten/v2/examples/rotate@latest)

2023/11/15 06:07:16 VersionUnavailable: GLX: Failed to create context: GLXBadFBConfig

I am wondering if this is an issue with the version of OpenGL? Can anyone point me in the right direction?

glxinfo shows the following output (truncated for brevity)
name of display: :0
display: :0  screen: 0
direct rendering: Yes
server glx vendor string: SGI
server glx version string: 1.4
server glx extensions:
    GLX_ARB_context_flush_control, GLX_ARB_create_context,
    GLX_ARB_create_context_no_error, GLX_ARB_create_context_profile,
    GLX_ARB_fbconfig_float, GLX_ARB_framebuffer_sRGB, GLX_ARB_multisample,
    GLX_EXT_create_context_es2_profile, GLX_EXT_create_context_es_profile,
    GLX_EXT_fbconfig_packed_float, GLX_EXT_framebuffer_sRGB,
    GLX_EXT_get_drawable_type, GLX_EXT_libglvnd, GLX_EXT_no_config_context,
    GLX_EXT_texture_from_pixmap, GLX_EXT_visual_info, GLX_EXT_visual_rating,
    GLX_MESA_copy_sub_buffer, GLX_OML_swap_method, GLX_SGIS_multisample,
    GLX_SGIX_fbconfig, GLX_SGIX_pbuffer, GLX_SGIX_visual_select_group,
    GLX_SGI_make_current_read
client glx vendor string: Mesa Project and SGI
client glx version string: 1.4

**Comment by SolarLune** [comment]
No idea, but here's some guesses. 

A quick search shows that an actual port to Pi would give better results, but it should (?) be possible to at least get it working: #760 

Ebitengine also uses GLFW to handle rendering and input, [so it might be good to check out how to get that to work first](https://forums.raspberrypi.com/viewtopic.php?t=222264)?

**Reply by irai** [reply]
Thanks @SolarLune. I have been reading enough on the topic to get totally confused :-)

I have the latest, fresh PI OS 64bit (Bookworm) install on a RPI3 with golang 1.21. I believe Bookworm uses GLFW as the underlying rendering lib so in theory GLFW should be working (i will need to confirm this). 

My next step is to try a previous OS to see if I can it working and work from there. Will post here when I find a result.

Would love to hear from others than got ebitengine working on a PI.
Cheers

**Reply by irai** [reply]
Update. 
I tried with a fresh bullsyeye OS install and got the same error when trying to run the rotate demo. 
The error implies that I need a certain version of GLX  but I am still none the wiser. It is the same error on the latest Bookworm OS too.

I am wondering if Ebitengine simply cannot run on the pi?

Here is the error I got:
> $ go run github.com/hajimehoshi/ebiten/v2/examples/rotate@latest
2023/11/20 11:06:07 VersionUnavailable: GLX: Failed to create context: GLXBadFBConfig
exit status 1

I have a fresh install of Raspberry OS Bullseye, fresh install of go 1.21 and openGL version 2.1.
> $ uname -am
Linux rpi3-bullseye 6.1.21-v7+ #1642 SMP Mon Apr  3 17:20:52 BST 2023 armv7l GNU/Linux

> $ go version
go version go1.21.4 linux/arm

OpenGL version:
> $ glxinfo | grep -e "OpenGL version" -e "Core" -e "Compatible"
OpenGL version string: 2.1 Mesa 20.3.5

I also managed to compile GLFW from sources and run the examples without any issues. 
Any ideas of what else I could try?

**Reply by SolarLune** [reply]
Maybe try a lower version of Ebitengine? Maybe 2.5 would work for you?

**Reply by irai** [reply, marked as answer]
HI @SolarLune , I tried several versions of ebitengine and several OS images and finally had success with Ubuntu server 20.04.05 LTS (32 bits) and ebitengine v2.6.3. 



Anyway, for anyone interested in a working version of ebitengine on RPI3, the following steps should give you a working ebitengine:
You will need a Raspberry PI 3 and SD card
- Use the Raspberry PI imager to install Ubuntu Server 20.04.05 LTS (32 bits) on the SD card.
- Boot the RPI3
- Download and install golang [go1.21.4.linux-armv6l.tar.gz](https://go.dev/dl/go1.21.4.linux-armv6l.tar.gz). 
- Install the ebitengine dependencies as per https://ebitengine.org/en/documents/install.html?os=linux
- install xterm and xinit - sudo apt install xterm xinit
- on the console, start X server with "startx xterm"
- go run github.com/hajimehoshi/ebiten/v2/examples/rotate@v2.6.3

That is. You should have a running rotating example on the Raspberry PI. 

Hope that is useful to others. 

ps: I could not make ebitengine work on any of the Raspberry PI OS images - suspect it has to do with the versions of Xorg/Wayland but requires someone with a better X11 knowledge than I to help.

**Comment by mcuadros** [comment]
The solution is just forcing the OpenGL version: 

`export MESA_GL_VERSION_OVERRIDE=4.5`

At least this worked for me on raspiban bookworm

**Reply by fipso** [reply]
This makes it work on my PI5 had the exact same error on my PI3 B+ before, so i guess this should also fix it there

**Comment by joelschutz** [comment]
A thing that I discovered are some docker images to compile games for arm64.

```shell
docker pull monkeyx/retro_builder:arm64
docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
docker run --privileged -it --platform=linux/arm64 --name builder64 --volume <your-game-root-path>:<where-to-mount-in-container> monkeyx/retro_builder:arm64 bash
```

These already come with all dependencies for ebitengine except Go, but you can add a binary to the mounted volume(it must be an arm version). Managed to compile for arm64 using the `build -buildvcs=false`, don't know exactly what that means. Still trying to configure my RPi to run it.

Got this info here: https://portmaster.games/docker.html

**Comment by gregorosaurus** [comment]
I had this same issue.   Things worked fine on a pi 4 or greater.  Pi 3 was a no go.
